internal/oprf/improved_kkrt.go

```diff
@@ -0,0 +1,222 @@
+	m      int   // number of message tuples
+	drbg   int
+}
+
+// NewKKRT returns a KKRT OPRF
+// m: number of message tuples
+// k: width of OT extension binary matrix
+// baseOT: select which baseOT to use under the hood
+// ristretto: baseOT implemented using ristretto
+func newImprovedKKRT(m, baseOT, drbg int, ristretto bool) (OPRF, error) {
+	// send k columns of messages of length k/8 (64 bytes)
+	baseMsgLen := make([]int, k)
+	for i := range baseMsgLen {
+		baseMsgLen[i] = k / 8 // 64 bytes
+	}
+
+	ot, err := ot.NewBaseOT(baseOT, ristretto, k, curve, baseMsgLen, cipherMode)
+	if err != nil {
+		return imprvKKRT{}, err
+	}
+
+	return imprvKKRT{baseOT: ot, m: m, drbg: drbg}, nil
+}
+
+// Send returns the OPRF keys
+func (ext imprvKKRT) Send(rw io.ReadWriter) (keys Key, err error) {
+	// sample random 16 byte secret key for AES-128
+	sk := make([]byte, 16)
+	if _, err = rand.Read(sk); err != nil {
+		return keys, err
+	}
+
+	// send the secret key
+	if _, err := rw.Write(sk); err != nil {
+		return Key{}, err
+	}
+
+	// sample choice bits for baseOT
+	s := make([]byte, k/8)
+	if _, err = rand.Read(s); err != nil {
+		return keys, err
+	}
+
+	// act as receiver in baseOT to receive k x k seeds for the pseudorandom generator
+	seeds := make([][]byte, k)
+	if err = ext.baseOT.Receive(s, seeds, rw); err != nil {
+		return keys, err
+	}
+
+	// receive masked columns u
+	paddedLen := (ext.m + util.PadTill512(ext.m)) / 8
+	u := make([]byte, paddedLen)
+	q := make([][]byte, k)
+	h := blake3.New()
+	for col := range q {
+		if _, err = io.ReadFull(rw, u); err != nil {
+			return keys, err
+		}
+
+		q[col] = make([]byte, paddedLen)
+		err = crypto.PseudorandomGenerateWithBlake3XOF(q[col], seeds[col], h)
+		if err != nil {
+			return keys, err
+		}
+		// Binary AND of each byte in u with the test bit
+		// if bit is 1, we get whole row u to XOR with q[row]
+		// if bit is 0, we get a row of 0s which when XORed
+		// with q[row] just returns the same row so no need to do
+		// an operation
+		if util.BitSetInByte(s, col) {
+			err = util.ConcurrentBitOp(util.Xor, q[col], u)
+			if err != nil {
+				return Key{}, err
+			}
+		}
+	}
+	runtime.GC()
+	q = util.TransposeByteMatrix(q)[:ext.m]
+
+	// store oprf keys
+	aesBlock, err := aes.NewCipher(sk)
+	return Key{block: aesBlock, s: s, q: q}, err
+}
+
+// Receive returns the OPRF output on receiver's choice strings using OPRF keys
+func (ext imprvKKRT) Receive(choices *cuckoo.Cuckoo, rw io.ReadWriter) (encodings [cuckoo.Nhash]map[uint64]uint64, err error) {
+	if int(choices.Len()) != ext.m {
+		return encodings, ot.ErrBaseCountMissMatch
+	}
+
+	// receive AES-128 secret key
+	sk := make([]byte, 16)
+	if _, err = io.ReadFull(rw, sk); err != nil {
+		return encodings, err
+	}
+
+	// compute code word using pseudorandom code on choice string r in a separate thread
+	var pseudorandomChan = make(chan [][]byte)
+	var errChan = make(chan error, 1)
+	go func() {
+		defer close(errChan)
+		d := make([][]byte, ext.m)
```

Comment:
Here's one place where we could reduce memory allocations.

1. The easy way. We know that PseudorandomCodes always returns 64 bytes so we should allocate a slice of 64 bytes array (as suggested by @GeorgeKolampas-Optable). Then instead of allocating in PseudorandomCodes, we should pass a destination slice as a parameter (or a pointer to 64 byte array but that might be more annoying once we get to number 2.)
2. We could allocate a flat byte array of ext.m*64 length and index into it with slice notation but I'm not sure if that's actually better. I think that a slice of arrays probably also ends up with contiguous storage in memory without the hassle of having to manually index into it. What do you think @fsaintjacques?

Reply:
1. I've done a bare minimum implementation of this which you can see in this branch: https://github.com/Optable/match/tree/kkrt-arrays
Performance locally is nearly identical to the normal slice implementation to within error for 1m and 5m.

| Implementation | Size | Time  |
|-----------------|-----|-------|
| slice of slices     | 1 m  | 3.32s |
| slice of slices     | 1 m  | 3.31s |
| slice of slices     | 1 m  | 3.23s |
| slice of arrays     | 1 m  | 3.30s |
| slice of arrays     | 1 m  | 3.39s |
| slice of arrays     | 1 m  | 3.25s |
| slice of slices     | 5 m  | 18.85s |
| slice of slices     | 5 m  | 17.46s |
| slice of slices     | 5 m  | 17.58s |
| slice of arrays     | 5 m  | 17.59s |
| slice of arrays     | 5 m  | 17.02s |
| slice of arrays     | 5 m  | 17.24s |

This was a bare minimum effort. Doing this properly will require a moderately large refactor.

2. I tried this and memory usage increased due to the padding being interspersed in a contiguous slice. In addition, clarity of the code decreased greatly because indexing methods are unintuitive.